flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/ContinuousSplitPlannerImpl.java

```diff
@@ -0,0 +1,172 @@
+   *
+   * <li>{@link ContinuousEnumerationResult#splits()} should contain initial splits
+   * discovered from table scan for {@link StreamingStartingStrategy#TABLE_SCAN_THEN_INCREMENTAL}.
+   * For all other strategies, splits collection should be empty.
+   * <li>{@link ContinuousEnumerationResult#toPosition()} points to the starting position
+   * for the next incremental split discovery with exclusive behavior. Meaning files committed
+   * by the snapshot from the position in {@code ContinuousEnumerationResult} won't be included
+   * in the next incremental scan.
+   */
+  private ContinuousEnumerationResult discoverInitialSplits() {
+    Optional<Snapshot> startSnapshotOptional = getStartSnapshot(table, scanContext);
+    if (!startSnapshotOptional.isPresent()) {
+      return new ContinuousEnumerationResult(Collections.emptyList(), null,
+          new IcebergEnumeratorPosition(null, null));
+    }
+
+    Snapshot startSnapshot = startSnapshotOptional.get();
+    LOG.info("Get starting snapshot id {} based on strategy {}",
+        startSnapshot.snapshotId(), scanContext.startingStrategy());
+    List<IcebergSourceSplit> splits;
+    IcebergEnumeratorPosition toPosition;
+    if (scanContext.startingStrategy() == StreamingStartingStrategy.TABLE_SCAN_THEN_INCREMENTAL) {
+      // do a full table scan first
+      splits = FlinkSplitPlanner.planIcebergSourceSplits(table, scanContext, workerPool);
+      LOG.info("Discovered {} splits from initial full table scan with snapshot Id {}",
+          splits.size(), startSnapshot.snapshotId());
+      toPosition = new IcebergEnumeratorPosition(startSnapshot.snapshotId(), startSnapshot.timestampMillis());
+    } else {
+      splits = Collections.emptyList();
+      // Use parentId to achieve the inclusive behavior.
+      // Note that it is fine if parentId is null.
+      Long parentSnapshotId = startSnapshot.parentId();
+      Long parentSnapshotTimestampMs = null;
+      if (parentSnapshotId != null) {
+        parentSnapshotTimestampMs = table.snapshot(parentSnapshotId).timestampMillis();
+      }
+
+      toPosition = new IcebergEnumeratorPosition(parentSnapshotId, parentSnapshotTimestampMs);
+      LOG.info("Start incremental scan with start snapshot (inclusive): id = {}, timestamp = {}",
+          startSnapshot.snapshotId(), startSnapshot.timestampMillis());
+    }
+
+    return new ContinuousEnumerationResult(splits, null, toPosition);
+  }
+
+  /**
+   * Optional is used because table may be empty and has no snapshot
+   */
+  @VisibleForTesting
+  static Optional<Snapshot> getStartSnapshot(Table table, ScanContext scanContext) {
+    switch (scanContext.startingStrategy()) {
+      case TABLE_SCAN_THEN_INCREMENTAL:
+      case INCREMENTAL_FROM_LATEST_SNAPSHOT:
+        return Optional.ofNullable(table.currentSnapshot());
+      case INCREMENTAL_FROM_EARLIEST_SNAPSHOT:
+        return Optional.ofNullable(SnapshotUtil.oldestAncestor(table));
+      case INCREMENTAL_FROM_SNAPSHOT_ID:
```

Comment:
This appears to be always inclusive?

Reply:
Except for `TABLE_SCAN_THEN_INCREMENTAL`, all other starting strategy is inclusive from the starting snapshot.